### ISSUE TITLE
Dynamic require breaks WebPack build

### ISSUE THREAD

**Issue by swaechter** [typeorm/typeorm#262]
Since a few days I am playing around with NodeJS, TypeORM and WebPack. I use WebPack as a bundler engine to simplify the deployment process and ran into a problem.

Sadly, the dynamic require in PlatformTools.load is unable to work in a WebPack build. I hacked the PlatformTools.load method, so you are able to understand the output:

![screen1](https://cloud.githubusercontent.com/assets/5626755/22626853/e50951bc-ebb6-11e6-9416-d16b9194fb12.png)
![screen2](https://cloud.githubusercontent.com/assets/5626755/22626855/e82a61e2-ebb6-11e6-8ef9-41d72aeb0e2d.png)

Do you know how to fix this problem while still using webpack (For example move the require directly to the drivers or an addition in WebPack)? I created a repository with a sample project: https://github.com/swaechter/typeorm_webpack

PS: Coming from a Java background with Spring Boot and Hibernate, this project is really awesome! The usage of decorators is so much more convenient than defining a [second] model schema. Keep up the great work!

**Comment by swaechter** [comment]
As an addition: If you set exprContextCritical in the webpack.config.js to false, you will see the warnings in the PlatformTools:

![screen3](https://cloud.githubusercontent.com/assets/5626755/22627151/8a554fee-ebbd-11e6-860f-13a06a48bd4b.png)

**Comment by pleerock** [comment]
Hi, are you sharing the codebase between frontend and backend? If yes, then you should not allow webpack to pickup original typeorm package on the frontend. You should use `alias` in webpack configuration and map to [this](https://github.com/typeorm/typeorm/blob/master/extra/typeorm-model-shim.js) file.

**Comment by swaechter** [comment]
Hi! No the application has no frontend at all (At least for the moment), so I only work with backend related code. As a reason of that I guess the shim won't solve the problem?

As mentioned in the initial answer, you can find a reproducible example here: https://github.com/swaechter/typeorm_webpack

**Comment by pleerock** [comment]
Then what is the reason to use webpack on the backend?

**Comment by swaechter** [comment]
Sorry for my late answer, I was ill multiple times.

It's all about vendoring: With WebPack I am able to generate a single bundle that includes all dependencies. Based on the bundle, I can create a hash of it and build a small and simple Debian package (Keep it simple and stupid - KISS principle).

On the other hand if I compile the application without WebPack, I always have to copy around the node_modules directory or build the application on the production system. I found out, that this is the Node-ish way, but in my opinion it's a complete no go and can also lead to situations like the one with Azer Koçulu or if the NPM repository goes offline (Rarely happens).

Is this a feature that you are able & willing to implement or where you say someone should follow the Node approach? I don't know the codebase of TypeORM, but maybe it's possible to move the include to the database driver class and add a getter that returns the driver implementation?

**Comment by pleerock** [comment]
no what you are doing is completely wrong. Don't use webpack on the backend side. You simply don't need to make it because it wont work as expected. It's not KISS. Conversely you make it complicated by applying non standard and bad practices. Having everything in a single file is a bad practice and will lead you to lot of problems including references to not yet undefined objects. Don't do it.

**Comment by swaechter** [comment]
Ok, thank you. In that case I'll drop the bundler and deploy the application the Node way.

**Comment by Zuzon** [comment]
Please fix it. I'm trying to use TypeORM with nwjs. Support of webpack is very important.

**Comment by shilpi-verma-izap** [comment]
hi @pleerock, I need to create application with react as frontend bundled via webpack and i need to use backend with typeorm. i am unable to get it how i can use typeorm database instance  in my frontend to make database insertions etc. Do i need to follow some api structure or what? i am stuck right now.


I just checked the comments here and found you are talking something similar here . Please help if you have got any idea. I need to share code between frontend and backend.

**Comment by pleerock** [comment]
@shilpi-verma-izap I can only help you if you have a typorm issue or answer your specific questions, I can't afford myself to setup a whole application architecture for you, you'll need to make research and implement it on your own.

It really depend on your project structure, if you already have a backend and frontend then you just call backend api endpoints from the frontend, and backend controllers does the job - save your database sent from the frontend in the database.

**Comment by shilpi-verma-izap** [comment]
Thanks @pleerock for your reply. Can you help me how can i make api call to typeorm Database. and where should i have typeorm in my javascript application. 


1. I have bundled my frontend app via webpack. (frontend-bundle.js)
2. I have my typeorm js code in separate bundle (backend-bundle.js)

3. Now how can i create Database interation from my frontend app. Do i need to include both bundled files separately  in my index.html 

4. Or should i create apis in my backend-bundle and make api calls from frontend. And can you give any reference around same?

**Comment by pleerock** [comment]
@shilpi-verma-izap yes you should create an api and call typeorm in there, you don't need to bundle typeorm - it is backend thing, not frontend. First learn how to setup an express server and how to call express endpoints from the frontend, then you'll be able to integrate typeorm easily.

**Comment by TimonVS** [comment]
@pleerock I'm running into the same issue. Bundling the backend with Webpack may make no sense, but when building a big app using AWS Lambda it's almost essential to use Webpack for tree shaking dependencies (the bundle may otherwise get far too big for AWS Lambda). Is there a workaround for this problem?

**Comment by pleerock** [comment]
adding @daniel-lang

**Comment by daniel-lang** [comment]
@TimonVS what is your issue? I'm under the impression that, because of #1143 webpack could be used for backend projects as well. Did something break in the meantime?

**Comment by TimonVS** [comment]
@daniel-lang You're right. I figured out that this issue was actually not the issue I was experiencing. I was running into problems with having a glob for entities and subscribers. After compiling the code with Webpack TypeORM wasn't able to find the classes to get the metadata from, I was getting this error:

```
2018-01-09T11:11:53.783Z	e407a8d8-f52d-11e7-9509-3db5a7e45e94	{ Error: Cannot find module "."
at webpackMissingModule (/var/task/src/api/v1/auth/forgot.js:27498:68)
at Function.module.exports.PlatformTools.load (/var/task/src/api/v1/auth/forgot.js:27498:146)
at /var/task/src/api/v1/auth/forgot.js:355351:61
at Array.reduce (native)
at Object.importClassesFromDirectories (/var/task/src/api/v1/auth/forgot.js:355350:32)
at ConnectionMetadataBuilder.module.exports.ConnectionMetadataBuilder.buildEntityMetadatas (/var/task/src/api/v1/auth/forgot.js:355310:86)
at Connection.module.exports.Connection.buildMetadatas (/var/task/src/api/v1/auth/forgot.js:138011:57)
at Connection.<anonymous> (/var/task/src/api/v1/auth/forgot.js:137671:30)
at step (/var/task/src/api/v1/auth/forgot.js:137546:23)
at Object.next (/var/task/src/api/v1/auth/forgot.js:137527:53)
at fulfilled (/var/task/src/api/v1/auth/forgot.js:137518:58)
at propagateAslWrapper (/var/task/src/api/v1/auth/forgot.js:195545:23)
at /var/task/src/api/v1/auth/forgot.js:195909:31
at /var/task/src/api/v1/auth/forgot.js:195582:70
at /var/task/src/api/v1/auth/forgot.js:195909:31
at process._tickDomainCallback [as _tickCallback] (internal/process/next_tick.js:135:7) code: 'MODULE_NOT_FOUND' }
```
I mitigated the issue by directly importing the entities and subscribers in ormconfig.js. But it cost me several hours and digging through TypeORM code to figure out the solution.

**Comment by daniel-lang** [comment]
I'm not sure I understand what you want to do. If you still have a problem and don't want to work with direct imports, please provide more information or upload your repo so I can take a look at it

**Comment by onionhammer** [comment]
Yeah tree shaking is the difference between a 10mb lambda and a 700kb lambda. This should really be fixed in the typeorm repo, as it is responsible for crashing when it is 'webpacked'

**Comment by pleerock** [comment]
>  I think this is a legitimate use for "webpack on the backend". (re @pleerock)

yes, sorry for my statement above, in some cases it is legitimate. I just meant some crazy things people do that I face when they mix webpack frontend backend without understanding responsibility of each technology.

Can somebody recap what are the exactly problems with typeorm + webpack currently?

**Comment by pleerock** [comment]
but its not an issue. Its an opportunity to use it this way. Just specify migrations explicitly: `migrations: [Migration1, Migration2, ...]`

**Comment by onionhammer** [comment]
@pleerock Thats the only issue I was seeing, and it was caused by a dynamic require of 'glob', which led me to this issue. I had to webpack ignore "pg-native" as well, but now my typeorm queries are currently running from a lambda after being webpacked.

A workaround is documented here, but I have not tested it:
https://github.com/typeorm/typeorm/blob/master/docs/faq.md#how-to-use-webpack-for-the-backend

I think it's better just to not do migrations in production (running webpack) anyway, better to do it as part of the CI/CD process and just use ts-node, so excluding migrations is fine for me.
